### PR TITLE
VocabBuilder.koplugin: Better handling word context

### LINKED ISSUE #11754: Nested dictionary lookup uses original sentence instead of using respective sentence from dictionary
* KOReader version: 2024.04
* Device: Kobo sage

#### Issue
When you save a word in vocab builder, it saves the word in context of its original sentence. So if you look up a word from a book, everything works as expected.

However, you can still do dictionary lookups _within_ dictionary's lookup. The problem is that in this case, the wrong sentence is used as context: it uses the original book sentence, rather than the dictionary one

I rely on these sentences for context so it's an important feature, especially since I'm using Anki's SRS to come back to these sentences+words multiple times

#### Steps to reproduce

1. Do a dictionary lookup from any book
2. Do another lookup from the dictionary definition popup
3. Add the word to the vocab builder
4. Close all popups, go to Vocabulary builder from the main menu, find the newly added word.
5. Press on the 3 dots which will show the word in the context of the sentence
6. Observe that the word from the dictionary lookup was used (from the 2nd step) but the original sentence was used (from the 1st step)

https://github.com/koreader/koreader/assets/6875790/6759b9d8-1a2c-4c1e-ac3d-b123baaa82c3

### LINKED ISSUE #12465: FR: Both Word stem & "word form in context" for Vocabulary Builder
First of all, thank you very much for this wonderful plugin (Vocabulary Builder).  This used to be the "killer feature" of Kindle.  Just a small problem though.  Koreader's vocabulary builder seems to store  whatever the dictionary query is, before combining the previous and next part of the context to form the whole.  This works quite well with English,  where fuzzy search can match different forms of a word (migrate, migrated, migrating etc.) However, when I try to look up korean words,  more often than not, the user has to tweak the query to get a hit.  For non-Korean speakers I will use English as an example. "He migrated to America and got married last year".  If this sentence is rendered in Korean and the user tries to look up "migrated", often there is not a single hit.  So the user has to manually change "migrated" to "migrate" ( The korean original form of a verb and its past tense in the middle of a sentence differ too much for fuzzy search to find a match). If the user chooses to add to VB in the modified lookup, The context would be "He migrate to America and got married last year", with the wrong form and tense. I also hope to export the words to Anki for future review, so getting the original form is necessary because of the need to fill the definition field.

**Possible solution**
A quick and fairly easy solution would be to take the approach of  Kindle's VB where both original word(word stem) and word form in the sentence are stored along with the context sentence.

Word stem: migrate
Word in context: migrated
Context: He migrated to America and got married last year.

When user has to manually tweak the query and do another lookup, the VB should store the second look up as word stem.

This will also be useful even for English words. Because if words are to be exported to Anki, their stems(original forms) will be needed to query the definitions that form the back of a card.

I don't know if I have made myself clear. If it is necessary, I can provide screenshots to clarify.

@weijiuqiao

### PR DESCRIPTION
**Pull request by weijiuqiao**
This pr retains the highlighted word in the vocab context when the word stem is being searched manually.
This does not fix #11754 but makes things a bit better.

Closes #12465.

- - -
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/12469)

### CODE REVIEW

File: plugins/vocabbuilder.koplugin/main.lua [hunk 1]
```diff
@@ -63,6 +63,19 @@ local function saveSettings()
...
+    text = text:gsub("^[\n%s]*", "")
+    text = text:gsub("[\n%s]*$", "")
+    -- Trim spaces around newlines
+    text = text:gsub("%s*\n%s*", "\n")
+    -- Trim consecutive spaces (that would probably have collapsed
+    -- in rendered CreDocuments)
+    text = text:gsub("%s%s+", " ")
+    return text
+end
+
```
Comment (NiLuJe) [inline plugins/vocabbuilder.koplugin/main.lua:78]: Possibly worth making the ReaderHighlight one public and use it instead of duplicating code?

Comment (weijiuqiao) [reply plugins/vocabbuilder.koplugin/main.lua:78]: Do I just add a line like `ReaderHighlight.cleanupSelectedText = cleanupSelectedText` or change the function to be under ReaderHighlight and also where it's being called?

Comment (hius07) [reply plugins/vocabbuilder.koplugin/main.lua:78]: https://github.com/koreader/koreader/blob/802a9068705f4d1564fc24bd2412eb1dfdfaf994/frontend/languagesupport.lua#L153-L155

Comment (NiLuJe) [reply plugins/vocabbuilder.koplugin/main.lua:78]: Ah, good catch! Then, yeah, probably move it to util and update every call site ;).

Comment (weijiuqiao) [reply plugins/vocabbuilder.koplugin/main.lua:78]: Do I do it or you guys do it? I'm not that confident in messing with the core codebase...

> Ah, good catch! Then, yeah, probably move it to util and update every call site ;).

Comment (NiLuJe) [reply plugins/vocabbuilder.koplugin/main.lua:78]: Done in https://github.com/koreader/koreader/pull/12477 ;).

**Comment by weijiuqiao** [comment]
Updated code to use `util.cleanupSelectedText`.

**Review by poire-z — APPROVED** [review]

**Pull request merged**